src/Chapter3.hs

```diff
@@ -971,6 +1082,31 @@ implement the following functions:
 🕯 HINT: to implement this task, derive some standard typeclasses
 -}
 
+data Weekday
+  = Monday
+  | Tuesday
+  | Wednesday
+  | Thursday
+  | Friday
+  | Saturday
+  | Sunday
+  deriving (Show, Eq, Ord, Enum, Bounded)
+
+isWeekend :: Weekday -> Bool
+isWeekend weekday =
+  case weekday of
+    Saturday -> False
+    Sunday -> False
+    _ -> True
+
+nextDay :: Weekday -> Weekday
+nextDay weekday
+  | weekday == maxBound = minBound
+  | otherwise = succ weekday
+
+daysToParty :: Weekday -> Int
+daysToParty weekday = (fromEnum Friday - fromEnum weekday) `mod` (maxBound + 1)
```

Comment:
I'm not sure this function returns a valid result :thinking: 

It looks like the `maxBound` variable has type `Int`, and for `Int` the value is a bit bigger than you expect, probably:

```haskell
λ: maxBound :: Int
9223372036854775807
λ: maxBound + 1 :: Int
-9223372036854775808
```

This happens sometimes when using polymorphic code and type inference. The types might not be what you expect... We had a bug in production because of this once, so it's often useful to specify types of polymorphic values in complicated expressions for clarity and maintainability :slightly_smiling_face:

Reply:
I was thinking on being parametric, but I guess there is still no reason to think about a week that doesn't have 7 days :joy: I'll fix that!

Reply:
I changed it to be `7`, but if I wanted to keep using the approach from before, I should then write `maxBound :: Weekday`, right?

Reply:
@g-belmonte Yes, using :seven: is the correct and simplest approach here :+1:  

Unfortunately, you can't use `maxBound :: Weekday` here. The `mod` function requires values to implement the `Integral` typeclass:

```haskell
λ: :t mod
mod :: Integral a => a -> a -> a
```

But for `Weekday` you only have `Enum` and `Bounded`. `Integral` is for numbers that can be divided. It is not derivable automatically, and, frankly speaking, the `Weekday` data type shouldn't implement this instance...

If you want to keep the implementation generic (the number of days in a week probably won't change soon, but for some enumerations that change often you may want to keep this property), you need to use 

```haskell
length [minBound :: Weekday .. maxBound]
```

The `[minBound .. maxBound]` list creates a list of all entries, and then you just call `length` on it. I guess, `fromEnum (maxBound :: Weekday) - fromEnum (minBound :: Weekday) + 1` should work as well.

Nevertheless, the pattern of creating all values with `[minBound .. maxBound]` is a very common pattern, and there are standard functions to capture it:

* https://hackage.haskell.org/package/relude-0.7.0.0/docs/Relude-Extra-Enum.html#v:universe